src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs

```diff
@@ -157,5 +157,29 @@ internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnota
             return this;
         }
 
+        /// <summary>
+        /// For scenarios such as overriding with differing ref kinds (such as out vs in or ref)
+        /// we need to compare function pointer parameters assuming that Ref matches RefReadonly/In
+        /// and Out. This is done because you cannot overload on ref vs out vs in in regular method
+        /// signatures, and we are disallowing similar overloads in source with function pointers.
+        /// </summary>
+        internal static bool RefKindEquals(TypeCompareKind compareKind, RefKind refKind1, RefKind refKind2)
+            => (compareKind & TypeCompareKind.FunctionPointerRefMatchesOutInRefReadonly) != 0
+               ? (refKind1 == RefKind.None) == (refKind2 == RefKind.None)
+               : refKind1 == refKind2;
+
+        /// <summary>
+        /// For scenarios such as overriding with differing ref kinds (such as out vs in or ref)
+        /// we need to compare function pointer parameters assuming that Ref matches RefReadonly/In
+        /// and Out. For that reason, we must also ensure that GetHashCode returns equal hashcodes
+        /// for types that only differ by the type of ref they have.
+        /// </summary>
+        internal static RefKind GetRefKindForHashCode(RefKind refKind)
+            => refKind switch
```

Comment:
>refKind switch [](start = 15, length = 14)

I think it would be more robust and clearer to base the hash code on the value ```(refKind == RefKind.None)```. There would be clear correspondence to the equality evaluation, consistency between implementation of both APIs would be obvious and wouldn't require any additional maintenance when the set of RefKinds is extended.

Reply:
There is also an option to not adjust hash code based on ref kinds.

---
In reply to: [426986569](https://github.com/dotnet/roslyn/pull/44257#discussion_r426986569) [](ancestors = 426986569)